### ISSUE TITLE
Excessive memory usage and memory overrun (dorado 0.3.1) killing run.

### ISSUE THREAD

**Issue by myxotheles** [nanoporetech/dorado#270]
Hi, I am re-posting this as a new issue, as I am not sure the comment to [https://github.com/nanoporetech/dorado/issues/240](url) was noticed.

When running dorado v0.3.1 with --modified-bases 5mCG_5hmCG, dorado uses excessive amount of memory. Regardless how much I allocate, it overruns. I have tried with up to 178GB RAM on 16 cores and 2 GPU (Ampere) but it crashed:

```
[2023-06-27 15:46:04.301] [info] > Creating basecall pipeline
[2023-06-27 16:20:50.595] [info] > Index parameters input by user: kmer size=15 and window size=10.
/opt/sge_spool/8.6.12b/sbg5/job_scripts/2991154: line 27: 196237 Bus error 
```

This seems to only occur with --modified-bases. Is this a known issue or am I missing something?

**Comment by tijyojwad** [comment]
Hi @myxotheles - thanks for reporting. We've had other reports as well, but unfortunately we haven't pushed a proper fix yet. We're looking into this. Thank you for your patience!

In the meantime, you may be able to work around this by splitting your dataset into smaller pod5s (or set of fewer pod5s). This is of course no ideal, but it may unblock you in the meantime.

**Comment by worromot** [comment]
Hi @tijyojwad - I have a similar issue. As you suggested, I tried running dorado v0.3.1 with --modified-bases 5mCG_5hmCG on a set of small pod5 files rather than a large merged file.
Still have the memory issue on 128G RAM allocated, after 3 hrs:
dorado basecaller $MODEL $INDIR --modified-bases 5mCG_5hmCG > $OUT  
[info] > Creating basecall pipeline                            
srun: Exceeded job memory limit] 3% [02h:59m:36s<03d:49m:43s]

**Comment by tijyojwad** [comment]
That's an interesting data point! Based on debugging so far I'd have expected the same memory issue regardless of fast5/pod5. How much smaller is your small set of pod5s compared to the original set? This is of course not ideal but perhaps going smaller may help. 128GB is a lot of memory, so I'm surprised even the smaller dataset is running out after just 3% processed (how many reads would that be?)

**Comment by worromot** [comment]
> ... How much smaller is your small set of pod5s compared to the original set? This is of course not ideal but perhaps going smaller may help. 128GB is a lot of memory, so I'm surprised even the smaller dataset is running out after just 3% processed (how many reads would that be?)

These are all the same data*, starting with ~2400 fast5 files with 4000 records each (so, 9.6M reads). They are either converted to a single pod5 file (~450GB in size) or using `--one-to-one` conversion to a separate dir with 2400 pod5 files.
Both ways of running pod5s resulted in out of memory error at about the same timepoint; 3% * 9.6M reads =~288K reads. 
Running fast5s is slower, but at this point it is at ~550K processed reads.

*I incorrectly read your suggestion and thought that running multiple individual pod5 files would be less taxing on memory that running the same files merged. Splitting the pod5 file set into 50 chunks should work.

**Comment by tijyojwad** [comment]
Yes sorry I could have been clearer - I did indeed mean splitting into chunks of smaller number of pod5s and the calling them in batches.

**Comment by worromot** [comment]
@StuartAbercrombie -- to clarify: unlike myxoteles, I have only tried modified calling, so out of memory issues are based on that.

**Comment by StuartAbercrombie** [comment]
Thanks for the clarification.  If anyone has a reliable repro. of excessive memory usage then a few things would help us debug it:

1) Information on whether it is specific to mod-base calling -- i.e. for a given dataset + settings, it only happens when mod-base calling is enabled.
2) Some indication of the memory growth over time (e.g. the output of top for the process at a few times during the run).
3) Whether the memory limit exceeded is the size of resident pages or the virtual memory.  (Pod5 files are read via memory mapping, which means a large virtual memory size accompanies large pod5 files, but the OS should kick out pages as necessary when the system needs it.)
4) Output of nvidia-smi and cat /proc/cpuinfo.
5) Whether it only happens with particular (large?) datasets.

**Comment by StuartAbercrombie** [comment]
There is a known limitation where, if output is stalled, e.g. if writes are going to a slow file system, then this back pressure could cause unlimited growth in a queue specific to mod base calling.  I have a change that would prevent this.  It would be interesting to know whether, in a reliable repro. case, directing output to /dev/null changes whether OOM occurs.

**Comment by worromot** [comment]
@StuartAbercrombie -- interesting. Let me see if outputting to /dev/null also leads to OOM

**Comment by worromot** [comment]
@StuartAbercrombie 
Tried outputting to /dev/null (on the same data)
Still OOM error (with 64G memory resrved):
`srun: Exceeded job memory limit] 2% [01h:14m:24s<02d:01h:08m:06s]`

**Comment by StuartAbercrombie** [comment]
That's interesting.  We have a fix in 0.3.2 for the unbounded queue, so it's perhaps worth trying that to check no other factor, beyond the file output, is causing queue growth
